### DISCUSSION TITLE
Can (should) I use Open WebUI as a front-end to my own (custom) HTTP server endpoints (not an OpenAI-compatible API)?

### DISCUSSION THREAD

**Discussion by maspotts** [open-webui/open-webui#5129] [Q&A]
Hi: we've developed an internal-use RAG/multimodal chat-with-our-documents service, fully dockerized, but we need a better web client: it looks as if Open WebUI is probably ideal, but all of the deployment docs I've come across are focused on calling either Ollama or an OpenAI-compatible API directly, whereas our (HTTP) servers provide our own (higher-level of abstraction) endpoints (and handles the complexity of RAG + LLM access itself): can I nevertheless use Open WebUI as a front-end for our own service, via its custom endpoints?

**Comment by thiswillbeyourgithub** [comment]
Not part of the team but I too have made [my own RAG tool](https://github.com/thiswillbeyourgithub/WDoc) and I'm hopeful that the pipelines allow for this. If you haven't take a look a their pipelines repo, it has rag examples. Also look for "action" in the official docs

**Comment by Peter-De-Ath** [comment]
Yes.

You can use the pipe functionality, You can create a pipe that you'll be able to expose your service as a model/s.
In there you can essentially call your own endpoints.
Using a pipe you can do whatever you want in there as long as you return the correct format.

The pipelines repo is a bit behind feature wise being able to utilise status updates/citations.

**Reply by maspotts** [reply]
Thanks! So using pipes, I can call my own server's endpoints? Or am I limited to only openAI-supported endpoints?

**Reply by thiswillbeyourgithub** [reply]
A pipe is essentially replacing the call to the llm so you can use any endpoint

**Reply by Peter-De-Ath** [reply]
No you're not limited, do whatever you want in the function, Doesn't even have to be AI/LLM related.

Have a look at this example, just uses the requests library to make a call.
https://openwebui.com/f/justinrahb/perplexity

There are plenty of other examples on the Open WebUI Community site, you should be able to piece together what you're wanting.
https://openwebui.com/functions/

**Reply by maspotts** [reply]
Great!  So it looks like I can simply create a pipe that knows how to query my service: do I need to conform to an OpenAI response format in order to get my server's output into the chat window?  Or can I customize that to consume my own output format?

**Reply by justinh-rahb** [reply]
Yes the output returned back to the WebUI needs to be OpenAI-ish, but the Pipe will allow you to call your own endpoint and translate as required.

**Reply by maspotts** [reply]
Thanks: i'll have a play around with that and see if I can make it work!

**Reply by maspotts** [reply]
It looks like pipelines is itself a server which conforms to the OpenAI API: so I guess maybe a simpler solution is just to modify my own server to give it an endpoint that conforms to the openAI API?  Can I then configure open Web eye to call my server as a local (OpenAI compatible) service directly?

**Reply by justinh-rahb** [reply]
@maspotts that would certainly eliminate one of the moving parts and be more ideal.

**Comment by maspotts** [comment]
One last question: is there any way for me to create custom inputs/controls in the UI?  I need the user to be able to select filter constraints (for filtering Qdrant metadata) and select various control options: I'm hoping that that's something I can do out of the box?  Or would I have to fork the project and hack it to add controls to the chat UI?  (I couldn't find any reference to this in the docs so I"m assuming probably the latter?!)

**Reply by Peter-De-Ath** [reply]
You can use Vavles/UserValves

Here is an example implementation https://openwebui.com/f/hub/openai_manifold

Chat Controls (per chat)
![image](https://github.com/user-attachments/assets/b940d7c2-0443-405d-a7ff-8c1afa1c13cc)

From Workspace -> Functions (click the cog on your function)
![image](https://github.com/user-attachments/assets/d1353185-5313-4074-a619-404830b908c1)

**Comment by maspotts** [comment]
ah, great: thanks!  (for some reason I didn't find this in the docs)

**Comment by NielsRogge** [comment]
Hi,

This is possible (asked Cursor how to do it). Basically you need to set the following environment variables:

```bash
export ENABLE_OPENAI_API=true
export OPENAI_API_BASE_URL=http://localhost:8082/v1
```
After which you can run `open-webui serve`. Here, my custom HTTP endpoint is running at localhost:8082. I've made my endpoint OpenAI compatible but also handles all the RAG itself.

Moreover, I had to apply this suggestion to see the UI locally: https://github.com/open-webui/open-webui/discussions/8009#discussioncomment-11712394.

**Comment by YahyaRochdi** [comment]
@maspotts  Did you fix the problem ? I am too trying to see if I can turn my own apis to a open ai api endpoint compatible ?
